sdk/python/kfp/components/graph_component.py

```diff
@@ -85,3 +89,165 @@ def pipeline_spec(self) -> pipeline_spec_pb2.PipelineSpec:
 
     def execute(self, **kwargs):
         raise RuntimeError('Graph component has no local execution mode.')
+
+    def _dedupe_pipeline_spec(
+        self, pipeline_spec: pipeline_spec_pb2.PipelineSpec
+    ) -> pipeline_spec_pb2.PipelineSpec:
+        """removes duplicated component spec and executor specs caused by tasks
+        calling on the same components.
+
+        component specs are only deduped when the executor specs are the
+        same, this means factors like two tasks called on the same
+        component with differing resource specs would not be deduped,
+        deduping would still occur if the caller tasks differ by their
+        input specifications since this not affect the executor spec of
+        the components
+        """
+        from dataclasses import dataclass
+
+        @dataclass
+        class clones_data_structure:
+
+            def __init__(self, component_name, component_spec, executor_name,
```

Comment:
Dataclasses will take case of this for you. See [example](https://realpython.com/python-data-classes/#basic-data-classes).